Document modification: EBS data volume replacement (#216)
scalar-labs/scalar-terraform: docs/CassandraOperation.md

  resource_type             = "t3.large"
  resource_count            = 3
  resource_root_volume_size = 64

>   enable_data_volume        = true  

Reviewer: Please remove the trailing white spaces.

Author: Thank you for the review comments.
Unwanted whitespace removed.

* Also when adding a new node to a cluster it will slow down the system.

## Replace a Cassandra Node

There are basically three ways to replace a Cassandra node as follows.

1. Replace the node, but not the volume
2. Replace the volume of the node, but not the node
3. Replace both the node and the volume

This section explains what needs to be done for each case.

### Case 1: Replace the node, but not the volume

When a node (VM or instance) is not working properly for some reason but the volume is fully functioning, only the node should be replaced with a new one.
You can replace the node by tainting the VM or the instance and the volume attachment as described below.

> Note that, when you taint the volume attachment terraform will try to attach the same data or commit log volume to the new instance.

Reviewer 2: ```suggestion
Note that, when you taint the volume attachment, terraform will try to attach the same data or commit log volume to the new instance.
```

Author: Thank you for the review suggestions.
The document updated accordingly.

terraform apply
```

> #### Post Recovery Steps

Reviewer 2: ```suggestion
#### Post Terraform Steps
```
Please update the others.

Author: Thank you for the review suggestions.
The document updated accordingly.

After the Cassandra node is replaced, you will need to perform the following manual steps to get the node back in the cluster.

* Find the UUID of the `data` or `commitlog` volume and add it to `fstab`.

```console
ssh -F ssh.cfg cassandra-[].internal.scalar-labs.com
lsblk -p -P -d -o name,serial,UUID,SIZE
NAME="/dev/nvme2n1" SERIAL="vol018d1871d19da76f1" UUID="af767839-b23d-4d1f-8a19-debbcfd6413c" HCTL="" SIZE="1T"
...

mkdir /data
chown cassandra:cassandra /data
sudo /bin/bash -c "echo 'UUID=af767839-b23d-4d1f-8a19-debbcfd6413c /data xfs defaults,nofail 0 2' >> /etc/fstab"
sudo mount -a
```

* Remove the IP of the new node from seeds in `casssandra.yaml`
* Add `JVM_OPTS="$JVM_OPTS -Dcassandra.replace_address_first_boot=<dead node ip>"` to the bottom of `cassandra-env.sh`
* `sudo systemctl start cassandra`
* Repair with Reaper at some later point 

Please note that, if you replace a seed node (IP-A), you should replace IP-A from seeds with a newly created Cassandra node IP in all the Cassandra nodes.

> ### Case 2: Replace the volume of the node, but not the node

Reviewer: Isn't it needed to stop Cassandra before replacing the volume with terraform taint & apply?

Author: Cassandra node stops automatically when data volume is corrupted (delete the data volume).
So, the current steps are working fine without stopping the nodes manually.

When the volume is not functioning properly for some reason but the node is alive, only the volume should be replaced with a new one.
You can replace the volume by tainting the volume as described below.

> Note that this *will permanently delete data* on that volume and attach the recreated data or commit log volume to the already existing node. 

Reviewer 2: ```suggestion
Note that this *will permanently delete data* on that volume and attach the recreated data or commit log volume to the existing node. 
```

Author: Thank you for the review suggestions.
The document updated accordingly.

When you do this, it is recommended to recover data from backup.

#### Azure
```console
terraform taint "module.cassandra.azurerm_managed_disk.cassandra_data_volume[0]"

terraform apply
```

#### AWS
```console
terraform taint "module.cassandra.aws_ebs_volume.cassandra_data_volume[0]"

terraform apply
```

#### Post Terraform Steps
After the volume is replaced, you will need to perform the following manual steps to get the volume back in the node.

* Replace the UUID of the `data` or `commitlog` volume in `fstab`.

> ```console 

Reviewer: Please remove the trailing white space.

Author: Thank you for the review comments.
Unwanted whitespace removed.

ssh -F ssh.cfg cassandra-[].internal.scalar-labs.com

# Find the name of newly attached EBS volume
lsblk -p -P -d -o name,serial,SIZE
NAME="/dev/nvme2n1" SERIAL="vol018d1871d19da76f1" HCTL="" SIZE="1T"
...

# Create a file system in the newly created volume
sudo mkfs -t xfs <name of volume>

# Find the `UUID` of the data or commit log volume
lsblk -p -P -d -o name,serial,UUID,SIZE
NAME="/dev/nvme2n1" SERIAL="vol018d1871d19da76f1" UUID="af767839-b23d-4d1f-8a19-debbcfd6413c" HCTL="" SIZE="1T"
...

# Update the UUID of `/data` or `/commitlog` directory 
sudo vi /etc/fstab

sudo systemctl daemon-reload

sudo mount -a

# Change ownership of `/data` directory 
sudo chown cassandra:cassandra /data
```
* Restore data from Cassy
* `sudo systemctl start cassandra`
* Repair with Reaper at some later point 

### Case 3: Replace both the node and the volume

When both a node and the volume are not working properly, both resources should be replaced with new ones.
You can do it by tainting both resources as described below.

> Since the volume is replaced as the 2nd option, this *will permanently delete data* on that volume.

Reviewer 2: ```suggestion
Since the volume is replaced in the same manner as the 2nd option, the data on that volume will be permanently deleted.
```

Author: Thank you for the review suggestions.
The document updated accordingly.

* Find the `instance` or `vm` you wish to replace and copy the line.

> * You also need to decide to taint either the *drive attachment* or the *volume* directly.

Author: Thank you for the review comment.
The document updated accordingly.

> 

Author: Thank you for the review comment.
The document updated accordingly.

> ### Case 1: Taint volume attachment

Reviewer 3: `Case 2 and Case 3` uses `Taint Volume`.
`Taint volume` or `Taint Volume` Which is correct?

---
📝 

Case 1 : Replace `VM` only
Case 2: Replace `VM` and `Volume`
Case 3: Replace `Volume` only

Author: @tei-k san,
Thank you for the detailed checking.

> Taint volume or Taint Volume Which is correct?

`Taint volume` is correct.
The document updated accordingly.

> ### Case 1: Taint volume attachment

Author: Thank you for the review comment.
The document updated accordingly.

> When you taint the volume attachment terraform will try to attach the same data or commit log volume to the new instance. This is the ideal situation as it is the quickest way to replace a node.

Reviewer 2: ```suggestion
When a node (VM or instance) is not working properly for some reason but the volume is fully functioning, only the node should be replaced with a new one.
You can replace the node by tainting the VM or the instance and the volume attachment as described below.
Note that, when you taint the volume attachment terraform will try to attach the same data or commit log volume to the new instance. 
```

Author: Thank you for the review comment.
The document updated accordingly.

terraform apply
```
#### Post Recovery Steps
After the Cassandra node is replaced, you will need to perform the following manual steps to get the node back in the cluster.

* Find the UUID of the `data` or `commitlog` volume and add it to `fstab`.

```console
ssh -F ssh.cfg cassandra-[].internal.scalar-labs.com
lsblk -p -P -d -o name,serial,UUID,SIZE
NAME="/dev/nvme2n1" SERIAL="vol018d1871d19da76f1" UUID="af767839-b23d-4d1f-8a19-debbcfd6413c" HCTL="" SIZE="1T"
...

mkdir /data
chown cassandra:cassandra /data
sudo /bin/bash -c "echo 'UUID=af767839-b23d-4d1f-8a19-debbcfd6413c /data xfs defaults,nofail 0 2' >> /etc/fstab"
sudo mount -a
```

* Remove the IP of the new node from seeds in `casssandra.yaml`
* Add `JVM_OPTS="$JVM_OPTS -Dcassandra.replace_address_first_boot=<dead node ip>"` to the bottom of `cassandra-env.sh`
* `sudo systemctl start cassandra`
* Repair with Reaper at some later point 

Please note that, if you replace a seed node (IP-A), you should replace IP-A from seeds with a newly created Cassandra node IP in all the Cassandra nodes.

> ### Case 2: Taint volume with node replacement

Reviewer 2: ```suggestion
### Case 3: Replace both a node and the volume
```
Also, this should come at the end.

Author: Thank you for the review comment.
The document updated accordingly.

> The other option is to taint the volume which should be used as a last resort. This *will permanently delete data* on that volume and it will recreate the node along with data or commit log volume. Be sure you can recover the data from a backup first.

Reviewer 2: ```suggestion
When both a node and the volume are not working properly, both resources should be replaced with new ones.
You can do it by tainting both resources as described below.
Since the volume is replaced as the 2nd option, this *will permanently delete data* on that volume.
```

Author: Thank you for the review comment.
The document updated accordingly.

> ```

Reviewer 2: The updated document got too many branches so it's pretty difficult to properly follow.
I think it needs to be reorganized to include a new case.
```
* Case 1
** Replace a resource
** Restart Cassandra
* Case 2 
** Replace a resource
** Restart Cassandra
...
```

Author: @feeblefakie san,
Thank you for the suggestion.
The document updated accordingly.

### Taint Volume

> The other option is to taint the volume which should be used as a last resort. This *will permanently delete data* on that volume. Be sure you can recover the data from a backup first.

Reviewer 2: I think you changed the context.
The original doc says `taint attachment` or `taint volume`. 
If it is removed, users can't follow it any more.

Author: Sorry, Orginal doc also says [Taint Volume](https://github.com/scalar-labs/scalar-terraform/blob/master/docs/CassandraOperation.md#taint-volume)

> There are two other options for replacing the taint volume as a last resort.

Reviewer 2: What do you mean `replacing the taint volume`?

Author: Sorry for the inconvenience.
Here I mean, replace the corrupted volume with the `taint` command.

Document modified.

#### Replacing the node along with taint volume replacement

This *will permanently delete data* on that volume and it will recreate the node along with data or commit log volume. Be sure you can recover the data from a backup first.

##### Azure
```console
terraform taint "module.cassandra.module.cassandra_cluster.azurerm_virtual_machine.vm-linux[0]"
terraform taint "module.cassandra.azurerm_managed_disk.cassandra_data_volume[0]"

terraform apply
```

##### AWS
```console
terraform taint "module.cassandra.module.cassandra_cluster.aws_instance.this[0]"
terraform taint "module.cassandra.aws_ebs_volume.cassandra_data_volume[0]"

terraform apply
```

#### Replacing the taint volume without node replacement

This *will permanently delete data* on that volume and attach the recreated data or commit log volume to the already existing instance. Be sure you can recover the data from a backup first.

> ##### Azure

Reviewer 2: Indenting is too deep. Should come up with a better way.

Author: The document updated accordingly.

terraform apply
```

### Taint Volume (without node replacement)

> The other option is to taint the volume which should be used as a last resort. This *will permanently delete data* on that volume and attach recreated volume with existing node. Be sure you can recover the data from a backup first.

Reviewer 3: In part `### Taint Volume (with node replacement)`,  it explicitly says `data volume`, what's the difference?

Author: > In part ### Taint Volume (with node replacement), it explicitly says data volume, what's the difference?

In part `Taint volume`, it handles both data or commitlog volume replacement.

Thank you for the review comments.
The document updated accordingly.

> The other option is to taint the volume which should be used as a last resort. This *will permanently delete data* on that volume and attach the recreated data or commit log volume to the already existing instance. Be sure you can recover the data from a backup first.

Author: Both options are true,
First option, replacing the instance along with taint volume.
Second option, only replacing taint volume. means it does not replace the node.

terraform apply
```

#### Post Recovery Steps
After the Cassandra node and volume are replaced, you will need to perform the following manual steps to get the node back in the cluster.

* Restore data from Cassy
  * You need to manually update the IP of the entry of `backup_history` table to match with the new IP
  * You need to manually update the IP address of Cassy backup with new IP in the cloud storage
* Remove the IP of the new node from seeds in `casssandra.yaml`
* Set `auto_bootstrap` to `false` in `cassandra.yaml`
    * this might not be needed. need to verify.
* Add `JVM_OPTS="$JVM_OPTS -Dcassandra.replace_address_first_boot=<dead node ip>"` to the bottom of `cassandra-env.sh`
* `sudo systemctl start cassandra`
* Repair with Reaper at some later point 

Please note that, if you replace a seed node (IP-A), you should replace IP-A from seeds with a newly created Cassandra node IP in all the Cassandra nodes.

> ### Case 3: Taint volume without node replacement

Reviewer 2: ```suggestion
### Case 2: Replace the volume of a node, but not the node
```

Author: Thank you for the review comment.
The document updated accordingly.

> Another option is to taint the volume which should be used as a last resort. This *will permanently delete data* on that volume and attach the recreated data or commit log volume to the already existing instance. Be sure you can recover the data from a backup first.

Reviewer 2: ```suggestion
When the volume is not functioning properly for some reason but the node is alive, only the volume should be replaced with a new one.
You can replace the volume by tainting the volume as described below.
Note that this *will permanently delete data* on that volume and attach the recreated data or commit log volume to the already existing node. When you do this, it is recommended to recover data from backup.
```

Author: Thank you for the review comment.
The document updated accordingly.

sudo mount -a
```

#### Replacing the data volume without node replacement
If you attached the new volume to the existing instance, you need to follow the these steps before starting Cassandra.

> * Replace the UUID of the `data` volume in `fstab`.

Reviewer 3: Is this just for data volume? Can we use the same procedure for commit volume?

Author: > Is this just for data volume? 

No

> Can we use the same procedure for commit volume?

Yes

Thank you for the review comments.
The document updated accordingly.
